### PR TITLE
feat: start L2 chain on Sepolia

### PR DESCRIPTION
**Pull request by lesterli**
## Summary

We plan to have a public demonet with L1 Sepolia, and firstly we would decouple the deployment script for the op finality gadget integration:

* start a private Babylon network or use Babylon public devnet
* start the L2 chain on Sepolia, explorer and birdge 
* start the finality gadget system

This PR is the first step to start the L2 chain on Sepolia:

* prepare for L2 chain
* launch L2 chain components: op-geth, op-node, op-proposer and op-batcher

Reference [op tutorials](https://docs.optimism.io/builders/chain-operators/tutorials/create-l2-rollup)

## Test Plan

Tested it on local L1 devnet.

1. start a local L1 chain, run the make target `start-local-l1-chain` to only start the L1's services `l1`, `l1-bn`, and `l1-vc`:
    ```

    cd deployments/finality-gadget-integration-op-l2
    make start-local-l1-chain
    ```
2. set the `.env` under the `deployments/finality-gadget-integration-op-l2/optimism`

    ```
    cd deployments/finality-gadget-integration-op-l2/optimism
    cp .env.example .env
    ```
    set L1 section with local devnet and use the pre-funded dev accounts as the admin, batcher, etc.
 
4. start the L2 chain under the `deployments/finality-gadget-integration-op-l2`

    ```
    cd deployments/finality-gadget-integration-op-l2
    export IMPL_SALT=$(openssl rand -hex 32)
    make start-op-chain-sepolia
    ```

### CODE REVIEW

**Review by bap2pecs — COMMENTED** [review]
- let's add a README.md file under the new /optimism folder to explain the "what" (what it is used for) and "how" (how it can be used)
- should `docker compose up -d l1 l1-bn l1-vc` be created as a Make target? so it can be used for local L1 deployment. also, is there another step to generate `/.devnet` before this command? (referenced in optimism/ops-bedrock/docker-compose.yml)

File: deployments/finality-gadget-integration-op-l2/optimism/generate-deploy-config.sh [hunk 1]
```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env bash
```
Comment (bap2pecs) [inline deployments/finality-gadget-integration-op-l2/optimism/generate-deploy-config.sh:1]: this is copied from optimism codebase. we should avoid

File: deployments/finality-gadget-integration-op-l2/Makefile [hunk 2]
```diff
@@ -104,4 +104,24 @@ stop: stop-op-devnet
 
 demo-test:
 	@./verify-babylon-finality.sh $(GIT_TOPLEVEL)/finality-gadget
-.PHONY: demo-test
\ No newline at end of file
+.PHONY: demo-test
+
+## Prepares for running a local OP chain
+pre-op-chain:
```
[the remark this answers is not in the document]
Comment (lesterli) [reply deployments/finality-gadget-integration-op-l2/Makefile]: The refactor is that these two targets are used to deploy the OP chain on the local L1 chain.

File: deployments/finality-gadget-integration-op-l2/optimism/install-deps.sh [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    else
+        echo "unsupported $(uname) system"
+        exit 1
+    fi
+    export PATH=$HOME/.foundry/bin:$PATH
+else
+    echo "Foundry is already installed"
+fi
+forge --version
+echo
```
[the remark this answers is not in the document]
Comment (lesterli) [reply deployments/finality-gadget-integration-op-l2/optimism/install-deps.sh:67]: Now the script `deployments/finality-gadget-integration-op-l2/prepare-op-devnet.sh` is removed and the script `deployments/finality-gadget-integration-op-l2/optimism/install-deps.sh` is used.

The `just install` is used when deploying the L1 contracts.

File: deployments/finality-gadget-integration-op-l2/optimism/README.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
+## Creating the OP chain testnet
+
+This deployment is used to create the OP chain testnet on the local L1 chain or the Sepolia testnet.
+
+TODO: update the make commands to the new structure.
+
+Deploy the OP chain testnet on the local L1 chain:
+
+```bash
+make start-op-devnet
```
Comment (bap2pecs) [inline deployments/finality-gadget-integration-op-l2/optimism/README.md:10]: I tested it out and there are a few issues:

1. we need to `cd` into the right repo
2. .env missing. there should be a `cp` step
3. I realized that running `install-deps.sh` is a huge pain. we should use containers for everything. the idea is we just need docker to be installed in any machine and no other dependencies needs to be instlaled

```
$ make start-op-devnet
OP_DIR: /Users/<redacted>/Documents/Projects/babylon-integration-deployment/optimism

Installing dependencies for OP L2...

just is already installed.
just 1.34.0

Installing Go...
--2024-10-05 18:56:02--  https://go.dev/dl/go1.22.6.linux-amd64.tar.gz
Resolving go.dev (go.dev)... 2001:4860:4802:36::15, 2001:4860:4802:32::15, 2001:4860:4802:34::15, ...
Connecting to go.dev (go.dev)|2001:4860:4802:36::15|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://dl.google.com/go/go1.22.6.linux-amd64.tar.gz [following]
--2024-10-05 18:56:02--  https://dl.google.com/go/go1.22.6.linux-amd64.tar.gz
Resolving dl.google.com (dl.google.com)... 2607:f8b0:4002:c10::5d, 2607:f8b0:4002:c10::5b, 2607:f8b0:4002:c10::88, ...
Connecting to dl.google.com (dl.google.com)|2607:f8b0:4002:c10::5d|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 68973343 (66M) [application/x-gzip]
Saving to: ‘STDOUT’

-                 0%[        ]       0  --.-KB/s               go/: Can't create 'go'
go/CONTRIBUTING.md: Failed to create dir 'go'
go/LICENSE: Failed to create dir 'go'
go/PATENTS: Failed to create dir 'go'
go/README.md: Failed to create dir 'go'
go/SECURITY.md: Failed to create dir 'go'
go/VERSION: Failed to create dir 'go'
go/api/: Failed to create dir 'go'
```

**Comment by bap2pecs** [comment]
`make start-local-l1-chain` depends on `pre-devnet` which has L2 related logic:

```
## Prepares for running a local devnet
pre-devnet: submodules $(DEVNET_CANNON_PRESTATE_FILES)
	@if ! [ -x "$(command -v geth)" ]; then \
		make install-geth; \
	fi
	@if ! [ -x "$(command -v eth2-testnet-genesis)" ]; then \
		make install-eth2-testnet-genesis; \
	fi
.PHONY: pre-devnet
```

this seems wrong to me. we shouldn't couple the "start L1" command with L2 logic.

File: deployments/finality-gadget-integration-op-l2/optimism/.env.example [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+GS_BATCHER_ADDRESS=0xa1AEF4C07AB21E39c37F05466b872094edcf9cB1
+GS_BATCHER_PRIVATE_KEY=0xe4d9cd91a3e53853b7ea0dad275efdb5173666720b1100866fb2d89757ca9c5a
+  
+# Proposer address
+GS_PROPOSER_ADDRESS=0x40E805e252D0Ee3D587b68736544dEfB419F351b
+GS_PROPOSER_PRIVATE_KEY=0x2d1f265683ebe37d960c67df03a378f79a7859038c6d634a61e40776d561f8a2
+  
+# Sequencer address
+GS_SEQUENCER_ADDRESS=0xC06566E8Ec6cF81B4B26376880dB620d83d50Dfb
+GS_SEQUENCER_PRIVATE_KEY=0x2a0290473f3838dbd083a5e17783e3cc33c905539c0121f9c76614dda8a38dca
```
Comment (bap2pecs) [inline deployments/finality-gadget-integration-op-l2/optimism/.env.example:31]: you mentioned in the test plan that we need to replace these keys in .env

> set L1 section with local devnet and use the pre-funded dev accounts as the admin, batcher, etc.

why not just use those keys here directly to save one step?

**Comment by bap2pecs** [comment]
test plan says to use `make start-op-chain-sepolia` to deploy on local L1. does this command work for both Sepolia and local L1. if so, should we rename this command?

**Comment by bap2pecs** [comment]
closing in favor of https://github.com/babylonlabs-io/babylon-integration-deployment/pull/26

**Pull request closed without merging**